### PR TITLE
Detect applications that are duplicates (modulo timestamps).

### PR DESCRIPTION
**Pull request by yotommy**
### Description

Detect applications that are duplicates (modulo timestamps).

## Release notes

This PR adds code to detect that a draft application being submitted contains identical answers to the currently-active application for this applicant and program. The scope of this PR is to log the condition; future PRs may affect the UX.

Though it would be cleanest to compare answers via their Java representations, these are not available at this point in the flow; only the JSON representation is available. Thus we compare the JSON representations of the applicant data, after clearing some timestamp metadata.

See [Design: Detecting Unintentional Duplicate Applications](https://docs.google.com/document/d/1nUGatPTxcL0bCuGY3e2hTajPAYufQJpXXnbpgXi4Xs4/edit?usp=sharing).

#### General

Read the full guidelines for PRs [here](https://docs.civiform.us/contributor-guide/developer-guide/technical-contribution-guide#guidelines-for-pull-requests)

- [x] Added the correct label: < feature | bug | dependencies | infrastructure | ignore-for-release | database >
- [x] Created unit and/or browser tests which fail without the change (if possible)
- [x] Performed manual testing (Chrome and Firefox if it includes front-end changes)
- [ ] Extended the README / documentation, if necessary

### Instructions for manual testing

As a logged-in user, edit an existing application. Do not change any answers before re-submitting.

As a guest user, complete an application. Then use the browser back button and re-submit.

In either case, check the server logs for `warn`-level messages that contain the string `"duplicate = true"`.

### CODE REVIEW

File: server/app/services/applicant/ApplicantData.java [hunk 1]
```diff
@@ -152,4 +153,29 @@ public Optional<LocalDate> getDateOfBirth() {
...
+
+    return thisApplicantData.asJsonString().equals(otherApplicantData.asJsonString());
+  }
+
+  private static void clearFieldsNotRequiredForComparison(ApplicantData applicantData) {
+    // The `updated_at` timestamp for an answer should not be considered when
+    // comparing answers.
+    try {
+      applicantData.getDocumentContext().set("$..updated_at", 0);
+    } catch (PathNotFoundException unused) {
```
Comment (yotommy) [inline server/app/services/applicant/ApplicantData.java:177]: There's an argument for being more defensive here and catching everything. Advice welcome.

File: server/app/services/applicant/ApplicantData.java [hunk 2]
```diff
@@ -152,4 +153,29 @@ public Optional<LocalDate> getDateOfBirth() {
...
+    clearFieldsNotRequiredForComparison(otherApplicantData);
+
+    return thisApplicantData.asJsonString().equals(otherApplicantData.asJsonString());
+  }
+
+  private static void clearFieldsNotRequiredForComparison(ApplicantData applicantData) {
+    // The `updated_at` timestamp for an answer should not be considered when
+    // comparing answers.
+    try {
+      applicantData.getDocumentContext().set("$..updated_at", 0);
```
Comment (bion) [inline server/app/services/applicant/ApplicantData.java:176]: **suggestion:** please include a comment explaining that the `..` scans the whole document

File: server/app/services/applicant/ApplicantData.java (hunk 1 above)
Comment (bion) [reply server/app/services/applicant/ApplicantData.java:177]: IMO this is fine

File: server/app/services/applicant/ApplicantData.java [hunk 3]
```diff
@@ -152,4 +153,29 @@ public Optional<LocalDate> getDateOfBirth() {
   public void setDateOfBirth(String dateOfBirth) {
     putDate(WellKnownPaths.APPLICANT_DOB, dateOfBirth);
   }
+
+  /**
+   * Returns `true` if the answers in `other` match the answers in the current object. Ignores
+   * certain metadata (such as timestamps) when comparing answers.
```
Comment (bion) [inline server/app/services/applicant/ApplicantData.java:159]: **suggestion (blocking):** recommend being a little more explicit:
```suggestion
   * Returns `true` if the answers in `other` match the answers in the current object. Ignores
   * `updated_at` timestamps when comparing answers.
```

**Review by bion — APPROVED** [review]

File: server/app/services/applicant/ApplicantData.java (hunk 2 above)
Comment (yotommy) [reply server/app/services/applicant/ApplicantData.java:177]: Done.

**Pull request merged**